momaland/envs/samegame/samegame.py

```diff
@@ -0,0 +1,531 @@
+tile at coordinates (x,y) is removed, this is encoded as the integer y * board_width + x.
+
+
+### Rewards
+
+Rewards can be team rewards or individual rewards (default is individual).
+If color_rewards = False:
+Dimension 0: n^2 points for the removal of any group of size n.
+If color_rewards = True (default):
+Dimensions d=0 to d=num_objectives-1: n^2 points for the removal of any group of size n in color d+1.
+
+### Version History
+
+"""
+
+# Adapted from https://github.com/waldner/samepy/tree/master
+# Copyright Davide Brini, 16/07/2014
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+
+from __future__ import annotations
+
+import copy
+import sys
+from typing_extensions import override
+
+import numpy as np
+from gymnasium import spaces
+from gymnasium.logger import warn
+from pettingzoo.utils import agent_selector, wrappers
+
+from momaland.utils.env import MOAECEnv
+
+
+def env(**kwargs):
+    """Returns the wrapped MOSameGame environment in `AEC` format.
+
+    Args:
+        **kwargs: keyword args to forward to the raw_env function
+
+    Returns:
+        A fully wrapped AEC env
+    """
+    env = raw_env(**kwargs)
+    # This wrapper terminates the game with the current player losing in case of illegal values.
+    # env = wrappers.TerminateIllegalWrapper(env, illegal_reward=-1)
+    # Asserts if the action given to step is outside the action space.
+    env = wrappers.AssertOutOfBoundsWrapper(env)
+    # Checks if function calls or attribute access are in a disallowed order.
+    # env = wrappers.OrderEnforcingWrapper(env)
+    return env
+
+
+def raw_env(**kwargs):
+    """Returns the MOBreakthrough environment in `AEC` format.
+
+    Args:
+        **kwargs: keyword args to forward to create the `MOBreakthrough` environment.
+
+    Returns:
+        A raw env.
+    """
+    return MOSameGame(**kwargs)
+
+
+class MOSameGame(MOAECEnv):
+    """SameGame environment for multiple agents with multiple objectives."""
+
+    metadata = {"render_modes": ["human"], "name": "mosamegame_v0", "is_parallelizable": False}
+
+    BLANK = 0
+
+    def __init__(
+        self,
+        board_width: int = 15,
+        board_height: int = 15,
+        num_colors: int = 5,
+        num_agents: int = 1,
+        team_rewards=False,
+        color_rewards=True,
+        render_mode=None,
+        seed=None,
```

Comment:
Seed should be assigned in reset no?

Reply:
Can we be sure that reset gets called before doing anything else? Setting the seed feels like a part of initialization to me.

Reply:
Yeah we can assume reset is called before anything else, that's part of the API flow.